internal/provider/data_source_vpc_router_route.go

```diff
@@ -0,0 +1,241 @@
+			},
+			"kind": schema.StringAttribute{
+				Computed:    true,
+				Description: "Whether the VPC router is custom or system created.",
+			},
+			"target": schema.SingleNestedAttribute{
+				Computed:    true,
+				Description: "location that matched packets should be forwarded to",
+				Attributes: map[string]schema.Attribute{
+					"type": schema.StringAttribute{
+						Description: "Route destination type. Possible values: vpc, subnet, instance, ip, internet_gateway, drop",
+						Computed:    true,
+					},
+					"value": schema.StringAttribute{
+						Description: "Depending on the type, it will be one of the following:" +
+							"- `vpc`: Name of the VPC " +
+							"- `subnet`: Name of the VPC subnet " +
+							"- `instance`: Name of the instance" +
+							"- `ip`: IP address " +
+							"- `internet_gateway`: name of the internet gateway",
+						Computed: true,
+					},
+				},
+			},
+			"timeouts": timeouts.Attributes(ctx),
+			"time_created": schema.StringAttribute{
+				Computed:    true,
+				Description: "Timestamp of when this VPC router route was created.",
+			},
+			"time_modified": schema.StringAttribute{
+				Computed:    true,
+				Description: "Timestamp of when this VPC router route was last modified.",
+			},
+		},
+	}
+}
+
+func (d *vpcRouterRouteDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var state vpcRouterRouteDataSourceModel
+
+	// Read Terraform configuration data into the model
+	resp.Diagnostics.Append(req.Config.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	readTimeout, diags := state.Timeouts.Read(ctx, defaultTimeout())
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	ctx, cancel := context.WithTimeout(ctx, readTimeout)
+	defer cancel()
+
+	params := oxide.VpcRouterRouteViewParams{
+		Route:   oxide.NameOrId(state.Name.ValueString()),
+		Project: oxide.NameOrId(state.ProjectName.ValueString()),
+		Router:  oxide.NameOrId(state.VPCRouterName.ValueString()),
+		Vpc:     oxide.NameOrId(state.VPCName.ValueString()),
+	}
+	route, err := d.client.VpcRouterRouteView(ctx, params)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Unable to read VPC router route:",
+			"API error: "+err.Error(),
+		)
+		return
+	}
+	tflog.Trace(ctx, fmt.Sprintf("read VPC router route with ID: %v", route.Id), map[string]any{"success": true})
+
+	// Parse vpcRouterRouteDestinationDataSourceModel into types.Object
+	dm := vpcRouterRouteDestinationDataSourceModel{
+		Type:  types.StringValue(string(route.Destination.Type)),
+		Value: types.StringValue(route.Destination.Value.(string)),
+	}
+	attributeTypes := map[string]attr.Type{
+		"type":  types.StringType,
+		"value": types.StringType,
+	}
+	destination, diags := types.ObjectValueFrom(ctx, attributeTypes, dm)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	state.Destination = destination
```

Comment:
I'm curious why you went with `types.Object` for `Destination` instead of a pointer to a `vpcRouterRouteDestinationDataSourceModel` like in the `oxide_image` data source. 

https://github.com/oxidecomputer/terraform-provider-oxide/blob/9f0e000cda86ea2c19b3b385c6f48acea84a95d8/internal/provider/data_source_image.go#L40

Then you could have skipped the parsing here. I'm not familiar with which way is technically superior but I figured I'd ask you about it.

Reply:
Frankly, while developing I got an error suggesting I use `types.Object` so I went with that